core/flyout.js

```diff
@@ -733,42 +851,75 @@ Blockly.Flyout.prototype.createBlockFunc_ = function(originBlock) {
 
 /**
  * Copy a block from the flyout to the workspace and position it correctly.
- * @param {!Blockly.Block} originBlock The flyout block to copy.
- * @param {!Blockly.Workspace} workspace The main workspace.
+ * @param {!Blockly.Block} originBlock The flyout block to copy..
  * @return {!Blockly.Block} The new block in the main workspace.
  * @private
  */
-Blockly.Flyout.prototype.placeNewBlock_ = function(originBlock, workspace) {
-  // Create the new block by cloning the block in the flyout (via XML).
-  var xml = Blockly.Xml.blockToDom(originBlock);
-  var block = Blockly.Xml.domToBlock(xml, workspace);
-  // Place it in the same spot as the flyout copy.
+Blockly.Flyout.prototype.placeNewBlock_ = function(originBlock) {
+  var targetWorkspace = this.targetWorkspace_;
   var svgRootOld = originBlock.getSvgRoot();
   if (!svgRootOld) {
     throw 'originBlock is not rendered.';
   }
-  var xyOld = Blockly.getSvgXY_(svgRootOld, workspace);
-  // Scale the scroll (getSvgXY_ did not do this).
-  if (this.RTL) {
-    var width = workspace.getMetrics().viewWidth - this.width_;
-    xyOld.x += width / workspace.scale - width;
-  } else {
-    xyOld.x += this.workspace_.scrollX / this.workspace_.scale -
-        this.workspace_.scrollX;
+  // Figure out where the original block is on the screen, relative to the upper
+  // left corner of the main workspace.
+  var xyOld = Blockly.getSvgXY_(svgRootOld, targetWorkspace);
+  // Take into account that the flyout might have been scrolled horizontally
+  // (separately from the main workspace).
+  // Generally a no-op in vertical mode but likely to happen in horizontal
+  // mode.
+  var scrollX = this.workspace_.scrollX;
+  var scale = this.workspace_.scale;
+  xyOld.x += scrollX / scale - scrollX;
```

Comment:
maybe just put xyOld.x += after the if (toolbox_at_right)?  Looks like the same calculation in both places once you've set scrollX and scale.
Also, why does the right vs left toolbox affect which workspace's scale you use? Shouldn't they both use the same one?